File: src/src_dotnet/JAStudio.Core/Note/BitFlagsSet.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace JAStudio.Core.Note;

public class BitFlagsSet : IEnumerable<int>
{
    private long _bitfield;

    public BitFlagsSet(long bitfield = 0)
    {
        _bitfield = bitfield;
    }

    public bool Contains(int value)
    {
        return (_bitfield & (1L << value)) != 0;
    }

    public bool ContainsBit(long value)
    {
        return (_bitfield & value) != 0;
    }

    public void SetFlag(int flag)
    {
        _bitfield |= (1L << flag);
    }

    public void UnsetFlag(int flag)
    {
        _bitfield &= ~(1L << flag);
    }

    public IEnumerator<int> GetEnumerator()
    {
        var bitfield = _bitfield;
        var flag = 0;
        while (bitfield != 0)
        {
            if ((bitfield & 1) != 0)
            {
                yield return flag;
            }
            bitfield >>= 1;
            flag++;
        }
    }

    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();

    public override int GetHashCode()
    {
        return _bitfield.GetHashCode();
    }

    public override bool Equals(object? obj)
    {
        if (obj is BitFlagsSet other)
        {
            return _bitfield == other._bitfield;
        }
        return false;
    }
}

File: src/jaspythonutils_src/jaspythonutils/sysutils/bit_flags_set.py
# from __future__ import annotations
#
# from typing import TYPE_CHECKING, override
#
# from typed_linq_collections.q_iterable import QIterable
#
# if TYPE_CHECKING:
#     from collections.abc import Iterator
#
# class BitFlagsSet(QIterable[int]):
#     __slots__: tuple[str, ...] = ("_bitfield",)
#
#     def __init__(self, bitfield: int = 0) -> None:
#         self._bitfield: int = bitfield
#
#     @override
#     def contains(self, value: int) -> bool:
#         return (self._bitfield & (1 << value)) != 0
#
#     def contains_bit(self, value: int) -> bool:
#         return (self._bitfield & value) != 0
#
#     def set_flag(self, flag: int) -> None:
#         self._bitfield |= (1 << flag)
#
#     def unset_flag(self, flag: int) -> None:
#         self._bitfield &= ~(1 << flag)
#
#     @override
#     def __iter__(self) -> Iterator[int]:
#         bitfield = self._bitfield
#         flag = 0
#         while bitfield:
#             if bitfield & 1:
#                 yield flag
#             bitfield >>= 1
#             flag += 1
#
#     @override
#     def __hash__(self) -> int:
#         return self._bitfield
#
#     @override
#     def __eq__(self, other: object) -> bool:
#         if isinstance(other, BitFlagsSet):
#             return self._bitfield == other._bitfield
#         return False
